File: exercises/D706/solution.pas
var
    q: array[1..10001] of longint;
    n, i, dummy, qs, qe: longint;
    s: string;
begin
    qs := 1;
    qe := 1;
    readln(n);
    for i := 1 to n do
    begin
        readln(s);
        if (copy(s, 1, 4) = 'PUSH') then
        begin
            val(copy(s, 6, 10), q[qe], dummy);
            inc(qe);
            if (qe = 10001) then
                qe := 0
        end
        else if (s = 'FRONT') then
        begin
            if (qs = qe) then
                writeln('Empty')
            else
                writeln(q[qs])
        end
        else if (s = 'POP') then
        begin
            if (qs = qe) then
                writeln('Cannot pop')
            else
            begin
                inc(qs);
                if (qs = 10001) then
                    qs := 0
            end
        end
        else
            writeln((10001 + qe - qs) mod 10001)
    end
end.

File: exercises/D706/solution.cpp
#include <iostream>
using namespace std;
int main() {
    int n, x, q[10001];
    string s;
    cin >> n;
    int qs = 0, qe = 0;
    for (int i = 0; i < n; i++) {
        cin >> s;
        if (s == "PUSH") {
            cin >> x;
            q[qe++] = x;
            if (qe == 10001) {
                qe = 0;
            }
        } else if (s == "FRONT") {
            if (qs == qe) {
                cout << "Empty\n";
            } else {
                cout << q[qs] << endl;
            }
        } else if (s == "POP") {
            if (qs == qe) {
                cout << "Cannot pop\n";
            } else {
                qs++;
                if (qs == 10001) {
                    qs = 0;
                }
            }
        } else {
            cout << (10001 + qe - qs) % 10001 << endl;
        }
    }
}
